Sources/_CShims/process_shims.c

```diff
@@ -0,0 +1,298 @@
+        &file_actions, &spawn_attr,
+        args, env
+    );
+    posix_spawn_file_actions_destroy(&file_actions);
+    posix_spawnattr_destroy(&spawn_attr);
+    return rc;
+}
+#endif // _POSIX_SPAWN
+
+int _subprocess_fork_exec(
+    pid_t * _Nonnull pid,
+    const char * _Nonnull exec_path,
+    const char * _Nullable working_directory,
+    const int file_descriptors[_Nonnull],
+    char * _Nullable const args[_Nonnull],
+    char * _Nullable const env[_Nullable],
+    uid_t * _Nullable uid,
+    gid_t * _Nullable gid,
+    int number_of_sgroups, const gid_t * _Nullable sgroups,
+    int create_session,
+    int create_process_group
+) {
+    int require_pre_fork = working_directory != NULL ||
+        uid != NULL ||
+        gid != NULL ||
+        (number_of_sgroups > 0 && sgroups != NULL) ||
+        create_session;
+
+#if _POSIX_SPAWN
+    // If posix_spawn is available on this platform and
+    // we do not require prefork, use posix_spawn if possible.
+    //
+    // (Glibc's posix_spawn does not support
+    // `POSIX_SPAWN_SETEXEC` therefore we have to keep
+    // using fork/exec if `require_pre_fork` is true.
+    if (require_pre_fork == 0) {
+        return _subprocess_posix_spawn_fallback(
+            pid, exec_path,
+            working_directory,
+            file_descriptors,
+            args, env,
+            create_process_group
+        );
+    }
+#endif
+
+    pid_t child_pid = fork();
+    if (child_pid != 0) {
+        *pid = child_pid;
+        return child_pid < 0 ? errno : 0;
+    }
+
+    if (working_directory != NULL) {
+        if (chdir(working_directory) != 0) {
+            return errno;
+        }
+    }
+
+    if (uid != NULL) {
+        if (setuid(*uid) != 0) {
+            return errno;
+        }
+    }
+
+    if (gid != NULL) {
+        if (setgid(*gid) != 0) {
+            return errno;
+        }
+    }
+
+    if (number_of_sgroups > 0 && sgroups != NULL) {
+        if (setgroups(number_of_sgroups, sgroups) != 0) {
+            return errno;
+        }
+    }
+
+    if (create_session != 0) {
+        (void)setsid();
+    }
+
+    if (create_process_group != 0) {
+        (void)setpgid(0, 0);
+    }
+
+    // Bind stdin, stdout, and stderr
+    int rc = dup2(file_descriptors[0], STDIN_FILENO);
+    if (rc != 0) { return rc; }
+    rc = dup2(file_descriptors[2], STDOUT_FILENO);
+    if (rc != 0) { return rc; }
+    rc = dup2(file_descriptors[4], STDERR_FILENO);
+    if (rc != 0) { return rc; }
+
+    // Close parent side
+    if (file_descriptors[1] != 0) {
+        rc = close(file_descriptors[1]);
+        if (rc != 0) { return rc; }
+    }
+    if (file_descriptors[3] != 0) {
+        rc = close(file_descriptors[3]);
+        if (rc != 0) { return rc; }
+    }
+    if (file_descriptors[4] != 0) {
+        rc = close(file_descriptors[5]);
+        if (rc != 0) { return rc; }
+    }
+
+    // Finally, exec
+    execve(exec_path, args, env);
+    // If we got here, something went wrong
```

Comment:
Indeed, how's the error code communicated back to the parent? I'd expect a pipe between parent and child here that has O_CLOEXEC. So if the parent reads something there, then that's the child's exit code. If the parent reads EOF, then everything's okay.

Here's code I've written in a previous life taking care of this: https://github.com/BromiumInc/BromiumCoreUtils/blob/2ec1c52a09831893618b9620383015e4362f7aa5/BromiumCoreUtils/BRUTask.m#L338-L341